server/src/main/java/org/elasticsearch/cluster/metadata/IndexReshardingState.java

```diff
@@ -0,0 +1,300 @@
+
+        public SourceShardState[] sourceShards() {
+            return sourceShards;
+        }
+
+        public TargetShardState[] targetShards() {
+            return targetShards;
+        }
+
+        /**
+         * Create resharding metadata representing a new split operation
+         * Split only supports updating an index to a multiple of its current shard count
+         * @param shardCount the number of shards in the index at the start of the operation
+         * @param multiple the new shard count is shardCount * multiple
+         * @return Split representing the start of the requested split
+         */
+        public static Split newSplitByMultiple(int shardCount, int multiple) {
+            assert multiple > 1 : "multiple must be greater than 1";
+
+            final int newShardCount = shardCount * multiple;
+            final var sourceShards = new SourceShardState[shardCount];
+            final var targetShards = new TargetShardState[newShardCount - shardCount];
+            Arrays.fill(sourceShards, SourceShardState.SOURCE);
+            Arrays.fill(targetShards, TargetShardState.CLONE);
+
+            return new Split(sourceShards, targetShards);
+        }
+
+        /**
+         * Get the current shard state of a source shard
+         * @param shardNum an index into the shards which must be no greater than the number of shards before split
+         * @return the source shard state of the shard identified by shardNum
+         */
+        public SourceShardState getSourceShardState(int shardNum) {
+            assert shardNum >= 0 && shardNum < sourceShards.length : "source shardNum is out of bounds";
+
+            return sourceShards[shardNum];
+        }
+
+        /**
+         * Set the current shard state of a source shard
+         * Currently the only legal transition is from SOURCE to DONE and any other transition will assert.
+         * This could be expressed through a markSourceDone API but this form is the same shape as {@link #setTargetShardState}
+         * and leaves the door open for additional source states.
+         * @param shardNum an index into the shards which must be no greater than the number of shards before split
+         * @param sourceShardState the state to which the shard should be set
+         */
+        public void setSourceShardState(int shardNum, SourceShardState sourceShardState) {
+            assert shardNum >= 0 && shardNum < sourceShards.length : "source shardNum is out of bounds";
+            assert sourceShards[shardNum].ordinal() + 1 == sourceShardState.ordinal() : "invalid source shard state transition";
+            assert sourceShardState == SourceShardState.DONE : "can only move source shard state to DONE";
+            for (var target : getTargetStatesFor(shardNum)) {
+                assert target == TargetShardState.DONE : "can only move source shard to DONE when all targets are DONE";
+            }
+
+            sourceShards[shardNum] = sourceShardState;
```

Comment:
We should not mutate these objects. This is way too dangerous, since the index-metadata is expected to be immutable. Instead we should add a `Builder` and use this to build a new `IndexReshardingState`.

Reply:
Thanks, I agree this is worth doing now. I should admit that I haven't really put a lot of thought into the client API for this state yet, since I thought that would be easier to get right when we were building the consumers. I expect it to churn until we actually have the state machine built into the resharding process.

Reply:
Done in 0e69a76 (but the API is still just enough to support a unit test for the moment)